File: src/models/evaluation/metrics.py
"""
Performance metrics calculation
"""
import numpy as np


def calculate_comprehensive_metrics(predictions, actuals, uncertainties):
    """
    Calculate comprehensive performance metrics

    Args:
        predictions: Predicted values
        actuals: Actual values
        uncertainties: Prediction uncertainties

    Returns:
        dict: Dictionary of metrics
    """
    # Basic regression metrics
    mse = np.mean((predictions - actuals) ** 2)
    rmse = np.sqrt(mse)
    mae = np.mean(np.abs(predictions - actuals))

    # Directional accuracy
    actual_direction = np.sign(actuals)
    pred_direction = np.sign(predictions)
    directional_accuracy = np.mean(actual_direction == pred_direction) * 100

    # Precision for up/down predictions
    up_mask = actual_direction > 0
    down_mask = actual_direction < 0

    up_precision = np.mean(pred_direction[up_mask] > 0) * 100 if up_mask.any() else 0
    down_precision = np.mean(pred_direction[down_mask] < 0) * 100 if down_mask.any() else 0

    # Hit rate for large movements
    large_moves = np.abs(actuals) > 0.02
    if large_moves.any():
        large_move_hit_rate = np.mean(
            np.sign(actuals[large_moves]) == np.sign(predictions[large_moves])
        ) * 100
    else:
        large_move_hit_rate = 0

    # Uncertainty quality (calibration)
    errors = np.abs(predictions - actuals)
    uncertainty_corr = np.corrcoef(errors, uncertainties)[0, 1] if len(errors) > 1 else 0

    # Sharpe ratio of trading signal
    signal_returns = np.where(
        pred_direction == actual_direction,
        np.abs(actuals),
        -np.abs(actuals)
    )
    sharpe_ratio = (
        np.mean(signal_returns) / np.std(signal_returns) * np.sqrt(252)
        if np.std(signal_returns) > 0 else 0
    )

    # Win rate
    win_rate = np.mean(signal_returns > 0) * 100

    metrics = {
        'MSE': mse,
        'RMSE': rmse,
        'MAE': mae,
        'Directional Accuracy': directional_accuracy,
        'Up Precision': up_precision,
        'Down Precision': down_precision,
        'Large Move Hit Rate': large_move_hit_rate,
        'Uncertainty Correlation': uncertainty_corr,
        'Sharpe Ratio': sharpe_ratio,
        'Win Rate': win_rate,
        'Avg Uncertainty': np.mean(uncertainties),
        'Max Uncertainty': np.max(uncertainties)
    }

    return metrics

File: cpp/src/models/metrics.cpp
// === FILE: cpp/src/models/metrics.cpp ===
#include "trading/models/metrics.hpp"

#include <cmath>

#include "trading/config/settings.hpp"

namespace trading::models {

namespace {
double sign(double x) { return (x > 0.0) ? 1.0 : (x < 0.0 ? -1.0 : 0.0); }
}  // namespace

Metrics compute_metrics(const std::vector<double>& predictions,
                        const std::vector<double>& actuals,
                        const std::vector<double>& uncertainties) {
    Metrics m;
    const std::size_t n = predictions.size();
    if (n == 0 || actuals.size() != n) {
        return m;
    }

    // RMSE / MAE.
    double sq_sum = 0.0;
    double abs_sum = 0.0;
    for (std::size_t i = 0; i < n; ++i) {
        const double err = predictions[i] - actuals[i];
        sq_sum += err * err;
        abs_sum += std::abs(err);
    }
    m.rmse = std::sqrt(sq_sum / static_cast<double>(n));
    m.mae = abs_sum / static_cast<double>(n);

    // Directional accuracy.
    std::size_t correct = 0;
    for (std::size_t i = 0; i < n; ++i) {
        if (sign(predictions[i]) == sign(actuals[i])) {
            ++correct;
        }
    }
    m.directional_accuracy = static_cast<double>(correct) / static_cast<double>(n) * 100.0;

    // Signal returns: +|actual| when direction matches, else -|actual|.
    std::vector<double> signal_returns(n, 0.0);
    double sr_mean = 0.0;
    std::size_t wins = 0;
    for (std::size_t i = 0; i < n; ++i) {
        const double sr = (sign(predictions[i]) == sign(actuals[i]))
                              ? std::abs(actuals[i])
                              : -std::abs(actuals[i]);
        signal_returns[i] = sr;
        sr_mean += sr;
        if (sr > 0.0) {
            ++wins;
        }
    }
    sr_mean /= static_cast<double>(n);

    double sr_var = 0.0;
    for (double sr : signal_returns) {
        const double d = sr - sr_mean;
        sr_var += d * d;
    }
    const double sr_std = std::sqrt(sr_var / static_cast<double>(n));

    m.sharpe_ratio = (sr_std > 0.0)
                         ? (sr_mean / sr_std) * std::sqrt(config::kAnnualizationDays)
                         : 0.0;
    m.win_rate = static_cast<double>(wins) / static_cast<double>(n) * 100.0;

    // Average uncertainty.
    if (!uncertainties.empty()) {
        double u_sum = 0.0;
        for (double u : uncertainties) {
            u_sum += u;
        }
        m.avg_uncertainty = u_sum / static_cast<double>(uncertainties.size());
    }

    return m;
}

}  // namespace trading::models
